src/MongoDB.Driver.Core/Core/Misc/ServerVersion.cs

```diff
@@ -0,0 +1,505 @@
+            return 0;
+
+            int ComparePreReleases()
+            {
+                if (_preRelease == null && other._preRelease == null)
+                {
+                    return 0;
+                }
+                else if (_preRelease == null)
+                {
+                    return 1;
+                }
+                else if (other._preRelease == null)
+                {
+                    return -1;
+                }
+
+                result = _preReleasePrefix.CompareTo(other._preReleasePrefix);
+                if (result != 0)
+                {
+                    return result;
+                }
+
+                if (_preReleaseNumericSuffix == null && other._preReleaseNumericSuffix == null)
+                {
+                    return 0;
+                }
+                else if (_preReleaseNumericSuffix == null)
+                {
+                    return -1;
+                }
+                else if (other._preReleaseNumericSuffix == null)
+                {
+                    return 1;
+                }
+
+                return _preReleaseNumericSuffix.Value.CompareTo(other._preReleaseNumericSuffix.Value);
+            }
+
+            int CompareCommitsAfterRelease()
+            {
+                if (_commitsAfterRelease == null && other._commitsAfterRelease == null)
+                {
+                    return 0;
+                }
+                else if (_commitsAfterRelease == null)
+                {
+                    return -1;
+                }
+                else if (other._commitsAfterRelease == null)
+                {
+                    return 1;
+                }
+
+                return _commitsAfterRelease.Value.CompareTo(other._commitsAfterRelease.Value);
+            }
+        }
+
+        /// <inheritdoc/>
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as ServerVersion);
+        }
+
+        /// <inheritdoc/>
+        public bool Equals(ServerVersion other)
+        {
+            return CompareTo(other) == 0;
+        }
+
+        /// <inheritdoc/>
+        public override int GetHashCode()
+        {
+            return ToString().GetHashCode();
+        }
+
+        /// <inheritdoc/>
+        public override string ToString()
+        {
+            var sb = new StringBuilder();
+            sb.Append(_major);
+            sb.Append('.');
+            sb.Append(_minor);
+            sb.Append('.');
+            sb.Append(_patch);
+            if (_preRelease != null)
+            {
+                sb.Append('-');
+                sb.Append(_preRelease);
+            }
+            if (_commitsAfterRelease != null)
+            {
+                sb.Append('-');
+                sb.Append(_commitsAfterRelease);
+            }
+            if (_commitHash != null)
+            {
+                sb.Append("-g");
+                sb.Append(_commitHash);
+            }
```

Comment:
Lines 311-325 could be shortened to:

    if (_preRelease != null)
    {
        sb.Append($"-{_preRelease}");
    }
    if (_commitsAfterRelease != null)
    {
        sb.Append($"-{_commitsAfterRelease}-g{_commitHash}");
    }